rbd: migration secret support

humblec opened the pull request:
rbd: parse migration secret and set it for CSI driver operations
    
    The intree secret has a data field called "key" which is the base64
    admin secret key. The ceph CSI driver currently expect the secret to
    contain data field "UserKey" for the equivalant. The CSI driver also
    expect the "UserID" field which is not available in the in-tree secret
    by deafult. This missing userID will be filled (if the username differ
    than 'admin') in the migration secret as 'adminId' field in the
    migration request, this commit adds the logic to parse this migration
    secret as below:
    
    "key" field value will be picked up from the migraion secret to "UserKey"
    field.
    "adminId" field value will be picked up from the migration secret to "UserID"
    field
    
    if `adminId` field is nil or not set, `UserID` field will be filled with
    default value ie `admin`.The above logic get activated only when the secret
    is a migration secret, otherwise skipped to the normal workflow as we have
    today.

Note:
E2E tests are added to make sure migration kind of secret is passed in tests.

Additional note:
The initial plan was to ask the admin to adjust the secret as a pre-upgrade step
for migration which avoid this change in the CSI driver, however taking out admin burden
is the main ask from kubernetes community so this change is introduced. None of the
existing workflow is broken for CSI driver with this change.

Updates #2509 


Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 commented:
@humblec this was not part of the original plan as per https://github.com/ceph/ceph-csi/issues/2509#issue-996778874 why we are adding this one? Do we have a clear design doc/plan about the changes required in cephcsi? IMHO we are adding more complexity to cephcsi to support migration.

humblec commented:
> @humblec this was not part of the original plan as per [#2509 (comment)](https://github.com/ceph/ceph-csi/issues/2509#issue-996778874) why we are adding this one? Do we have a clear design doc/plan about the changes required in cephcsi? IMHO we are adding more complexity to cephcsi to support migration.

@Madhu-1 I will add details about this  change in the PR description and commit ..etc, thats why I kept this in draft atm . I have already added this change in the tracker and can add some more details there as well which explain how this  helps. This looks to be pretty much the last change (unless some minor change comes up while doing complete e2e testing which cover kube.etc).  To give some context the initial plan was to ask this change as part of the pre upgrade step,  but  found a way to help admin to stay away from this change mostly and we can take care of it from csi driver code itself without breaking any existing functionality. Reg# documentation, it is there in the todo as well. Thanks :+1:

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec committed:
e2e: adjust deletion, filesystem and block  tests for migration volume

this commit create and make use of migration secret in the requests and
validate various csi operations

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 commented on e2e/ceph_user.go:
```diff
@@ -14,13 +14,17 @@ const (
 	keyringRBDNamespaceProvisionerUsername = "cephcsi-rbd-ns-provisioner"
 	keyringRBDNamespaceNodePluginUsername  = "cephcsi-rbd-ns-node"
+	keyringRBDMigrationProvisionerUsername = "cephcsi-rbd-mig-provisioner"
+	keyringRBDMigrationNodePluginUsername  = "cephcsi-rbd-mig-node"
```
do we need to create a new ceph user for this one? why not use existing users?

humblec replied:
we can , however this is the same pattern it has followed for radosnamespace test. so imo, its good to have a seperate user for this migration test alone.  wdyt?

humblec replied:
as discussed, will try to stick to the existing users for now .. addressing it.

humblec replied:
Done. thanks.

Madhu-1 commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+	}
+	if userName != "admin" {
+		sc.StringData["adminId"] = userName
+	}
```
what action we are taking if the username is admin?  i don't see its getting added to the secret template

humblec replied:
if its admin, no changes required, we just need to get the Key which is already part of the secret. Thats why this PR and to make the migration effort even smoother.

Madhu-1 commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+	}
+	if userKey == "" {
+		return fmt.Errorf("empty user key is passed")
+	}
```
IMO this validation is not required here.

Madhu-1 commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+	err = deleteResource(rbdExamplePath + "storageclass.yaml")
+	if err != nil {
+		return fmt.Errorf("failed to delete storageclass with error %w", err)
+	}
```
move this outside of this function as  this  function is a helper to create  the Secret and SC

humblec replied:
Done

Madhu-1 commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+	param["csi.storage.k8s.io/node-stage-secret-name"] = rbdMigrationNodePluginSecretName
+	err = createRBDStorageClass(f.ClientSet, f, scName, nil, param, deletePolicy)
+	if err != nil {
+		return fmt.Errorf("failed to create storageclass with error %w", err)
```
`failed to create storageclass with error %w` to `failed to create storageclass: %w`  we stopped adding `with  error` to the logs/error message

humblec replied:
Done.

Madhu-1 commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+	}
+
+	return nil
+}
```
IMO this block is not needed if we reuse the existing user and just create the secret and use it.

humblec replied:
as discussed the user key fetch is still required, so keeping it there for now.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -810,18 +810,27 @@ func (cs *ControllerServer) DeleteVolume(
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		// set req.Secrets
+		req.Secrets = passedSecretMap
```
IMHO this is becoming more ugly. we should not manipulate the input request and again assign the manipulated input to the request.

humblec replied:
I dont think so.  Again, we are just adjusting it in while we are in migration context, otherwise none of the code paths are touched. This is pretty shim change compared to the things we do here for other code paths, for ex:  chaning the entire nodestage request itself and intercpting the call and re issuing ...etc whlie we are in healer context .

nixpanic replied:
Changing the `*csi.DeleteVolumeRequest` that has been passed by the caller does not seem right. Better so something like

```go
secrets := req.GetSecrets()
if isMigrationSecret(secrets) {
    secrets, err = parseAndSetSecretMapFromMigSecret(secrets)
    ...
}

cr, err := util.NewUserCredentials(secrets)
```

humblec replied:
Done.thanks.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -837,16 +846,19 @@ func (cs *ControllerServer) DeleteVolume(
+		pErr = DeleteMigratedVolume(ctx, pmVolID, cr)
+		if pErr != nil && !errors.Is(pErr, ErrImageNotFound) {
+			return nil, status.Error(codes.Internal, pErr.Error())
 		}
```
why this change is required?

humblec replied:
It is just an adjustment or not really part of this the change this PR did, but good to have.  @Madhu-1 do I need to put it as a seperate PR or commit ?

humblec replied:
as discussed, added this logic to its own commit.. thanks.

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -32,6 +33,12 @@ func isMigrationVolID(volHash string) bool {
 }
 
+// isMigrationSecret validates if the passed in secretmap is a secret
+// of a migration volume request.
```
the comment doesn't actually explains what exactly we are doing it.

humblec replied:
sure .. will address it next refresh

humblec replied:
Done.

Madhu-1 reviewed (dismissed):
@humblec as its a migration work, why not ask the admin to recreate the secret and use it or take another route as we did for clusterID generation? IMO these changes make the cephcsi logic ugly to handle the migration.

humblec commented:
> @humblec as its a migration work, why not ask the admin to recreate the secret and use it or take another route as we did for clusterID generation? IMO these changes make the cephcsi logic ugly to handle the migration.

@Madhu-1 as mentioned in the other PR comment #2562 (comment), this was the initial thought ie let admin patch the secret. Then after some discussions with community folks , it was recommended that, take care of it maximum from the CSI side . So this PR, the advantage with this PR is that, admin Dont want to patch the secret at all, if the admin user is not different from admin which is mostly the case.

humblec commented:
@Madhu-1 all the comments are addressed as per the discussion, ptal.. Thanks :+1:

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented on internal/rbd/migration.go:
```diff
@@ -114,3 +123,31 @@ func parseAndDeleteMigratedVolume(ctx context.Context, volumeID string, cr *util
+	// this function can be called independently with a map of secret values
+	// it is good to have this check in place, also it gives clear error about this
+	// was hit on migration request compared to general one.
+	if len(passedSecretMap) != 0 {
```
if you really want to check for this error, return something on failure and have the happy-path of the function continue under the failure detection. This way you reduce indention, can add more failure detections easily and improve readability.

humblec replied:
Done.

nixpanic commented on internal/rbd/migration.go:
```diff
@@ -114,3 +123,31 @@ func parseAndDeleteMigratedVolume(ctx context.Context, volumeID string, cr *util
 	return nil
 }
+
+func parseAndSetSecretMapFromMigSecret(passedSecretMap map[string]string) (map[string]string, error) {
```
New functions should have a comment. Specially a function that does some parsing/conversion. The input and output needs to be documented.

humblec replied:
done.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -91,6 +91,10 @@ const (
+	// migration user name.
+	migUserName = "admin"
+	// migration user key.
+	migUserKey = "key"
```
this is also used as "key" in the map, can you use something that is more descriptive? it would surely help when printing the map in logs or such.

humblec replied:
unfortunately this is the field in the in the intree secret. so have to stay with it.

nixpanic commented on internal/util/credentials.go:
```diff
@@ -24,8 +24,8 @@ import (
 const (
-	credUserID           = "userID"
-	credUserKey          = "userKey"
+	CredUserID           = "userID"
```
this surely should be internal. If there really is a need for other functions to use this, move the function to this file.

humblec replied:
done.thanks

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -846,16 +846,19 @@ func (cs *ControllerServer) DeleteVolume(
 		return &csi.DeleteVolumeResponse{}, nil
 	}
-
 	rbdVol, err := genVolFromVolID(ctx, volumeID, cr, req.GetSecrets())
```
unneeded change, just keep the empty line

humblec replied:
done.

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -259,7 +259,14 @@ func (ns *NodeServer) NodeStageVolume(
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		// fill it in the request too as stageTransaction and other following also use req.Secrets()
+		req.Secrets = newSecretMap
```
don't do this, anywhere

humblec replied:
addressed

humblec committed:
rbd: parse migration secret and set it for controller server operations

This commit adds a couple of helper functions to parse the migration
request secret and set it for further csi driver operations.

More details:

The intree secret has a data field called "key" which is the base64
admin secret key. The ceph CSI driver currently expect the secret to
contain data field "UserKey" for the equivalant. The CSI driver also
expect the "UserID" field which is not available in the in-tree secret
by deafult. This missing userID will be filled (if the username differ
than 'admin') in the migration secret as 'adminId' field in the
migration request, this commit adds the logic to parse this migration
secret as below:

"key" field value will be picked up from the migraion secret to "UserKey"
field.
"adminId" field value will be picked up from the migration secret to "UserID"
field

if `adminId` field is nil or not set, `UserID` field will be filled with
default value ie `admin`.The above logic get activated only when the secret
is a migration secret, otherwise skipped to the normal workflow as we have
today.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

nixpanic commented on internal/util/credentials.go:
```diff
@@ -119,3 +122,30 @@ func GetMonValFromSecret(secrets map[string]string) (string, error) {
+
+// ParseAndSetSecretMapFromMigSecret parse the secretmap from the migration request and return
+// newsecretmap with the userID and userKey fields set.
+func ParseAndSetSecretMapFromMigSecret(secretmap map[string]string) (map[string]string, error) {
```
Is there a reason why there is no unit-test for this function?

humblec replied:
no specific reason, but added that too.

nixpanic commented on internal/rbd/migration.go:
```diff
@@ -19,19 +19,30 @@ package rbd
+// of a migration volume request. The migration secret carry a field
+// called `key` which is the equivalent of `userKey` which is what we
+// check here for identifying the secret.
+func isMigrationSecret(passedSecretMap map[string]string) bool {
```
Did I  not suggest to move this function to the `utils` package where the Credentials struct lives?

humblec replied:
actually not, but I moved in this refresh :)

nixpanic commented on internal/util/credentials.go:
```diff
@@ -119,3 +122,30 @@ func GetMonValFromSecret(secrets map[string]string) (string, error) {
+	// parse and set userKey
+	if secretmap[migUserKey] == "" {
+		return nil, fmt.Errorf("passed secret map does not contain user key")
+	}
```
Use the validation of `isMigrationSecret()`?

humblec replied:
done.

nixpanic commented on internal/rbd/migration.go:
```diff
@@ -85,15 +85,16 @@ func parseMigrationVolID(vh string) (*migrationVolID, error) {
+		log.ErrorLog(ctx, "migration volume details are nil")
 
-		return err
+		return fmt.Errorf("migration volume details are %v", parsedMigHandle)
```
these two lines are rather inconsistent... one explicitly reports `nil`, the other uses the variable (which is `nil`). Might be cleaner to create an error instance, log it and then return it.

humblec replied:
Done.

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -154,7 +154,7 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
 	ctx context.Context,
 	req *csi.NodeStageVolumeRequest,
-	cr *util.Credentials) (*rbdVolume, error) {
+	cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
```
I think the coding guidelines suggest to place this on a new line

nixpanic replied:
Is there really a need to pass `Credentials` and `secrets`? They would usually contain the same details, I think... If it is required, explain the reason in the comment above the function.

humblec replied:
contents are different, I have added the comment above the function too.

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -254,13 +254,20 @@ func populateRbdVol(
-	cr, err := util.NewUserCredentials(req.GetSecrets())
+	secrets := req.GetSecrets()
+	if isMigrationSecret(secrets) {
+		secrets, err = util.ParseAndSetSecretMapFromMigSecret(secrets)
```
`ParseAndSetSecretMapFromMigSecret()` could use `isMigrationSecret()` and this code gets a little nicer too.

humblec replied:
done,

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -353,19 +360,12 @@ func (ns *NodeServer) stageTransaction(
 	req *csi.NodeStageVolumeRequest,
 	volOptions *rbdVolume,
-	staticVol bool) (stageTransaction, error) {
+	staticVol bool, cr *util.Credentials) (stageTransaction, error) {
```
on a new line? maybe directly after `req` so that similar parameters are grouped together

humblec replied:
Done.

nixpanic commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+// set as the `adminId` field in the secret.
+func createRBDMigrationSecret(f *framework.Framework, secretName, userName, userKey string) error {
+	scPath := fmt.Sprintf("%s/%s", rbdExamplePath, "secret.yaml")
+	sc, err := getSecret(scPath)
```
`sc` usually refers to StorageClass, maybe use `sec` or `secret` instead?

humblec replied:
done.

nixpanic commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+// equivalent to the `UserKey` field, migration secret has `key` field.
+// if 'userName' has passed and if it is not admin, the passed in userName will be
+// set as the `adminId` field in the secret.
+func createRBDMigrationSecret(f *framework.Framework, secretName, userName, userKey string) error {
```
can you not use `createRBDSecret()` here?

humblec replied:
with the changes by introducing flags in args...etc , we can change, but considering there are many callers for createRBDSecret at this time, and also having a seperate function helps to make it clean, this is used atm.

nixpanic commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+}
+
+// createProvNodeCephUserAndSecret fetches the ceph migration user's key and create migration secret
+// with it based on the arg values of 'provSecret' and 'nodeSecret'.
```
After reading more of the test case, I now understand that `provSecret` refers to `provisioner`, not `provided`. Better spell out the whole name, or at least make it clear in the comment.

humblec replied:
Done. renamed to provisionersecret..

nixpanic commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+			rbdProvisionerCaps(defaultRBDPool, radosNamespace),
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create user %s with error %w", keyringRBDProvisionerUsername, err)
```
try not to introduce new error reporting that contains `with error`, that is unneeded long and as it is an error, rather unneeded. Use `"failed to create user %q: %w"` instead.

humblec replied:
Done.

nixpanic commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+		}
+		err = createRBDMigrationSecret(f, rbdMigrationProvisionerSecretName, keyringRBDProvisionerUsername, key)
+		if err != nil {
+			return fmt.Errorf("failed to create provisioner secret with error %w", err)
```
same here, and more below.

humblec replied:
Done.

nixpanic reviewed (dismissed):
Not much of functional comments, I guess. These are mostly related to maintainability and readability of the code.

humblec committed:
rbd: add unittests for IsMigrationSecret and ParseAndSetSecretMapFromMigSecret

This commit adds unit tests for newly introduced migration specific
functions.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: split the parsing and deletion logic to its own functions.

parseAndDeleteMigratedVolume() prviously clubbed the logic of
parsing of migration volume handle and then continued with the
deletion of the volume. however this commit split this
logic into two, ie parsing has been done in parseMigrationVolID()
and DeleteMigratedVolume() deletes the backend volume.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: parse migration secret and set fields for nodestage operations

this commit make use of the migration request secret parsing and set
the required fields for further nodestage operations

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
> Not much of functional comments, I guess. These are mostly related to maintainability and readability of the code.

Sure, addressed those comments too :+1:  ptal. thanks @nixpanic .

nixpanic commented on internal/util/credentials_test.go:
```diff
@@ -0,0 +1,85 @@
+package util
```
add copyright notice and license header please

humblec replied:
done.

nixpanic commented on internal/util/credentials.go:
```diff
@@ -119,3 +122,38 @@ func GetMonValFromSecret(secrets map[string]string) (string, error) {
+	// was hit on migration request compared to general one.
+	if len(secretmap) == 0 {
+		return nil, fmt.Errorf("passed secret is empty")
+	}
```
I still wonder why this is not part of `IsMigrationSecret()`.

humblec replied:
can be part of.. moved it .

nixpanic commented on internal/util/credentials.go:
```diff
@@ -31,6 +31,9 @@ const (
 	tmpKeyFileNamePrefix = "keyfile-"
+	migUserName          = "admin"
+	migUserID            = "adminId"
+	migUserKey           = "key"
```
as asked before, does this really need to be `key`, can it not be something specific for migration? Using `key` as key is a little confusing, it will make logs or debugging less easy to understand.

humblec replied:
unfortunately this has to be 'key', this is the field in existing intree secret which is untouched at time of migration , so we have to keep the same field.. thanks.

nixpanic replied:
ok, thanks. This could have been added as a comment to make it clearer. The commit message has some details, but ideally some of that can be found in the sources too.

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

nixpanic commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+			Secrets(cephCSINamespace).
+			Delete(context.TODO(), rbdMigrationProvisionerSecretName, metav1.DeleteOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to delete provisioner secret with error %w", err)
```
missed a couple of these `... with error: %v` strings

humblec replied:
adjusted these too.. ptal.. thanks.

nixpanic reviewed (commented):
Is there a design document for the tasks that are needed to be done, or is https://github.com/ceph/ceph-csi/projects/18 the only description that is available? It would be good to have a document that contains the steps the migration takes, with that we can validate that the code does what is required. This current PR looks close to complete, but I would like to know what the general plan is.

humblec commented:
>Is there a design document for the tasks that are needed to be done, or is https://github.com/ceph/ceph-csi/projects/18 the only description that is available?

The tracker issue ( https://github.com/ceph/ceph-csi/issues/2509) carry the changes identified based on the discussions so far. Unless something comes at e2e testing with kubernetes changes , thats pretty much done. 

> It would be good to have a document that contains the steps the migration takes, with that we can validate that the code does what is required. 

Indeed thats the plan, I have documentation in todo , also wrapping up the changes in kubernetes , so that e2e testing and examples ..etc can be captured to a doc.  and update here in this repo too.

nixpanic commented on e2e/rbd.go:
```diff
@@ -392,6 +394,15 @@ var _ = Describe("RBD", func() {
+
+				err = deleteProvNodeMigrationSecret(f, true, true)
+				if err != nil {
+					e2elog.Failf("failed to delete migration users and Secrets associated with error %v", err)
```
*sigh* more `with error` reporting. If this PR needs an update, please clean them up too.

humblec replied:
sure @nixpanic . Thanks!

Madhu-1 commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+	}
+	// if its admin, we dont need to change anything in the migration secret, the CSI driver
+	// will use the key from existing secret and continue.
+	if userName != "admin" {
+		sec.StringData["adminId"] = userName
+	}
```
if the user name is admin the secret will contain only the key?

humblec replied:
yes @Madhu-1

Madhu-1 commented on e2e/migration.go:
```diff
@@ -144,3 +144,133 @@ func generateClusterIDConfigMapForMigration(f *framework.Framework, c kubernetes
+
+// deleteProvNodeMigrationSecret deletes ceph migration secrets based on the
+// arg values of 'provisionerSecret' and 'nodeSecret'.
+func deleteProvNodeMigrationSecret(f *framework.Framework, provisionerSecret, nodeSecret bool) error {
```
do we need `provisionerSecret, nodeSecret` secret arguments here? i don't see apart from this testing no one is going to use this one

Madhu-1 commented:
@humblec please don't dismiss the requested review as am reviewing the PR. this can make the PR to be merged

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -1606,12 +1617,24 @@ var _ = Describe("RBD", func() {
 					e2elog.Failf("failed to generate clusterID configmap with error %v", err)
 				}
-				err = validateRBDStaticMigrationPV(f, appPath, false)
+				// create node user and migration secret.
+				err = createProvNodeCephUserAndSecret(f, false, true)
+				if err != nil {
+					e2elog.Failf("failed to create users and secret with error %v", err)
+				}
```
why create and delete and re-create the user in multiple tests? cant we create why in BeforeEach and Delete it in AfterEach?

humblec replied:
as discussed,  will do as a follow up PR along with `with error` changes.

Madhu-1 replied:
@humblec please add this to your TODO list

humblec replied:
sure.. added madhu..thansk

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -74,15 +74,16 @@ func parseMigrationVolID(vh string) (*migrationVolID, error) {
-	parsedMigHandle, err := parseMigrationVolID(volumeID)
-	if err != nil {
-		log.ErrorLog(ctx, "failed to parse migration volumeID: %s , err: %v", volumeID, err)
+func DeleteMigratedVolume(ctx context.Context, parsedMigHandle *migrationVolID, cr *util.Credentials) error {
```
no need to expose this function its internal

humblec replied:
sure.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -842,17 +851,21 @@ func (cs *ControllerServer) DeleteVolume(
+		if pErr != nil {
+			return nil, status.Error(codes.InvalidArgument, pErr.Error())
+		}
+		pErr = DeleteMigratedVolume(ctx, pmVolID, cr)
```
we are passing the pmVolID here but I don't see when parseMigrationVolID can return both error as nil and volID as nil?

humblec replied:
sure, the nil check was additional measure, however removed.

Madhu-1 commented on internal/util/credentials.go:
```diff
@@ -119,3 +122,35 @@ func GetMonValFromSecret(secrets map[string]string) (string, error) {
+	var err error
+	// parse and set userKey
+	if !IsMigrationSecret(secretmap) {
+		return nil, fmt.Errorf("passed secret map does not contain user key or it is nil")
```
use errors.New as formatting is not requested

humblec replied:
done.

Madhu-1 commented on internal/util/credentials.go:
```diff
@@ -119,3 +122,35 @@ func GetMonValFromSecret(secrets map[string]string) (string, error) {
+// newsecretmap with the userID and userKey fields set.
+func ParseAndSetSecretMapFromMigSecret(secretmap map[string]string) (map[string]string, error) {
+	newSecretMap := make(map[string]string)
+	var err error
```
why we need this err variable in this function as we are not using it instead just returning at end

humblec replied:
can be removed. done.

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
@Mergifyio rebase

humblec commented:
/retest all

humblec commented:
@Madhu-1 removed DNM on your behalf :+1:
